net/client.go

```diff
@@ -201,81 +215,89 @@ func (s *server) getRecords(ctx context.Context, id thread.ID, lid peer.ID, offs
 		Body: body,
 	}
 
-	lg, err := s.net.store.GetLog(id, lid)
+	logAddrs, err := s.net.store.Addrs(tid, lid)
 	if err != nil {
 		return nil, err
 	}
 
+	var (
+		recs   = newRecords()
+		wg     sync.WaitGroup
+		logErr = func(addr ma.Multiaddr, f func(addr ma.Multiaddr) error) {
+			if err := f(addr); err != nil {
+				log.Error(err.Error())
+			}
+		}
+	)
+
 	// Pull from each address
-	recs := newRecords()
-	wg := sync.WaitGroup{}
-	for _, addr := range lg.Addrs {
+	for _, addr := range logAddrs {
 		wg.Add(1)
-		go func(addr ma.Multiaddr) {
+
+		go logErr(addr, func(addr ma.Multiaddr) error {
 			defer wg.Done()
-			p, err := addr.ValueForProtocol(ma.P_P2P)
+			pid, ok, err := s.callablePeer(addr)
 			if err != nil {
-				log.Error(err)
-				return
-			}
-			pid, err := peer.Decode(p)
-			if err != nil {
-				log.Error(err)
-				return
-			}
-			if pid.String() == s.net.host.ID().String() {
-				return
+				return err
+			} else if !ok {
+				// skip calling itself
+				return nil
 			}
 
-			log.Debugf("getting records from %s...", p)
+			log.Debugf("getting records from %s...", pid)
 
 			client, err := s.dial(pid)
 			if err != nil {
-				log.Errorf("dial %s failed: %s", p, err)
-				return
+				return fmt.Errorf("dial %s failed: %w", pid, err)
 			}
+
 			cctx, cancel := context.WithTimeout(ctx, PullTimeout)
 			defer cancel()
 			reply, err := client.GetRecords(cctx, req)
 			if err != nil {
-				log.Warnf("get records from %s failed: %s", p, err)
-				return
+				log.Warnf("get records from %s failed: %s", pid, err)
+				return nil
 			}
+
 			for _, l := range reply.Logs {
-				log.Debugf("received %d records in log %s from %s", len(l.Records), l.LogID.ID, p)
+				var logID = l.LogID.ID
+				log.Debugf("received %d records in log %s from %s", len(l.Records), logID, pid)
+
+				if l.Log != nil && len(l.Log.Addrs) > 0 {
+					if err = s.net.store.AddAddrs(tid, logID, addrsFromProto(l.Log.Addrs), pstore.PermanentAddrTTL); err != nil {
```

Comment:
Mostly to avoid moving the head of the log. Method `getRecords` now just fetches the records and except adding new addresses and keys is stateless, real processing and state update happens in `putRecord`.